32-bit physics with FV3_RAP

SamuelTrahanNOAA opened the pull request:
- [x] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [x] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [x] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR
are specified below.

- [ ] ~~Results for one or more of the regression tests change and the reasons for the changes are understood and explained below.~~ **There are new regression tests, but no tests should change results. It is critical that all tests be run against their existing baselines, to ensure 64-bit physics does not change results on any platform.**

- [ ] ~~New or updated input data is required by this PR. If checked, please work with the code managers to update input data sets on all platforms.~~

## Description

This PR gets 32-bit physics working in FV3 within UFS. The ccpp portion of this is based on work done by NRL on the Neptune model, merged a few weeks ago. With much help from @DusanJovic-NOAA @grantfirl and @climbfuji, I've gotten it working in UFS as well. What works is:

1. All suites compile with 32-bit physics and either 64-bit or 32-bit dynamics.
2. The FV3_RAP and FV3_HRRR suites can run and produce output in either of those modes.
3. There are regression tests that ensure those don't break.
4. No existing tests change results.

For the purposes of testing, it is *critical* that all existing regression tests run against their current baselines, to ensure 64-bit physics do not change answers on any platforms. The only tests that need new baselines are the 32-bit physics tests, which I have copied to 32bit.conf and 32bit_gnu.conf for your convenience. (They are also in rt.conf and rt_gnu.conf.)

## Testing

How were these changes tested? What compilers / HPCs was it tested with? Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) Have regression tests and unit tests (utests) been run? On which platforms and with which compilers? (Note that unit tests can only be run on tier-1 platforms)

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel
- [x] cheyenne.gnu
- [x] gaea.intel
- [x] jet.intel
- [x] wcoss2.intel
- [ ] opnReqTest for newly added/changed feature
- [ ] CI

## Issues

Fixes https://github.com/ufs-community/ufs-weather-model/issues/1288

## Dependencies

https://github.com/noaa-psd/stochastic_physics/pull/59
https://github.com/NOAA-EMC/fv3atm/pull/534
https://github.com/NCAR/ccpp-physics/pull/930
https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/193
~~https://github.com/earth-system-radiation/rte-rrtmgp/pull/180~~ this was already merged

DusanJovic-NOAA commented:
I can not clone this branch from Hera:
```
$ git clone --recursive --branch sing_prec_from_main https://github.com/SamuelTrahanNOAA/ufs-weather-model
Cloning into 'ufs-weather-model'...
remote: Enumerating objects: 19882, done.
remote: Counting objects: 100% (69/69), done.
remote: Compressing objects: 100% (60/60), done.
remote: Total 19882 (delta 17), reused 14 (delta 9), pack-reused 19813
Receiving objects: 100% (19882/19882), 61.24 MiB | 28.03 MiB/s, done.
Resolving deltas: 100% (16007/16007), done.

 . . .

Cloning into '/scratch2/NCEPDEV/fv3-cam/Dusan.Jovic/ufs/32bit/ufs-weather-model/FV3'...
Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: clone of 'ssh://git@github.com/SamuelTrahanNOAA/fv3atm' into submodule path '/scratch2/NCEPDEV/fv3-cam/Dusan.Jovic/ufs/32bit/ufs-weather-model/FV3' failed
Failed to clone 'FV3' a second time, aborting
```

Please change submodule URL's to https.

SamuelTrahanNOAA commented:
> Please change submodule URL's to https

Done. The entire tree now uses https.

DusanJovic-NOAA commented:
> > Please change submodule URL's to https
> 
> Done. The entire tree now uses https.

Thanks.

climbfuji commented:
> > > Please change submodule URL's to https
> > 
> > 
> > Done. The entire tree now uses https.
> 
> Thanks.

Thanks. I had the same problem and had to switch to https locally as well.

DusanJovic-NOAA commented:
I made this changes in FV3:
```diff
diff --git a/ccpp/data/GFS_typedefs.F90 b/ccpp/data/GFS_typedefs.F90
index f125489..320f9cf 100644
--- a/ccpp/data/GFS_typedefs.F90
+++ b/ccpp/data/GFS_typedefs.F90
@@ -1,6 +1,6 @@
 module GFS_typedefs
 
-   use machine,                  only: kind_phys, kind_dbl_prec
+   use machine,                  only: kind_phys, kind_dbl_prec, kind_sngl_prec
    use physcons,                 only: con_cp, con_fvirt, con_g,                       &
                                        con_hvap, con_hfus, con_pi, con_rd, con_rv,     &
                                        con_t0c, con_cvap, con_cliq, con_eps, con_epsq, &
diff --git a/module_fcst_grid_comp.F90 b/module_fcst_grid_comp.F90
index 886f23a..a713fbe 100644
--- a/module_fcst_grid_comp.F90
+++ b/module_fcst_grid_comp.F90
@@ -38,6 +38,8 @@ if (rc /= ESMF_SUCCESS) write(0,*) 'rc=',rc,__FILE__,__LINE__; if(ESMF_LogFoundE
                                 atmos_model_exchange_phase_2,              &
                                 addLsmask2grid, atmos_model_get_nth_domain_info
 
+  use GFS_typedefs,       only: kind_phys, kind_sngl_prec
+
   use constants_mod,      only: constants_init
   use fms_mod,            only: error_mesg, fms_init, fms_end,             &
                                 write_version_number, uppercase
@@ -161,6 +163,7 @@ if (rc /= ESMF_SUCCESS) write(0,*) 'rc=',rc,__FILE__,__LINE__; if(ESMF_LogFoundE
     integer,dimension(2,6):: decomptile                  !define delayout for the 6 cubed-sphere tiles
     integer,dimension(2)  :: regdecomp                   !define delayout for the nest grid
     type(ESMF_Decomp_Flag):: decompflagPTile(2,6)
+    type(ESMF_TypeKind_Flag) :: grid_typekind
     character(3)          :: myGridStr
     type(ESMF_DistGrid)   :: distgrid
     type(ESMF_Array)      :: array
@@ -188,6 +191,12 @@ if (rc /= ESMF_SUCCESS) write(0,*) 'rc=',rc,__FILE__,__LINE__; if(ESMF_LogFoundE
     call ESMF_InfoGet(info, key="layout", values=layout, rc=rc); ESMF_ERR_ABORT(rc)
     if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
 
+    if (kind_phys == kind_sngl_prec) then
+      grid_typekind = ESMF_TYPEKIND_R4
+    else
+      grid_typekind = ESMF_TYPEKIND_R8
+    endif
+
     if (trim(name)=="global") then
       ! global domain
       call ESMF_InfoGet(info, key="tilesize", value=tilesize, rc=rc); ESMF_ERR_ABORT(rc)
@@ -200,6 +209,7 @@ if (rc /= ESMF_SUCCESS) write(0,*) 'rc=',rc,__FILE__,__LINE__; if(ESMF_LogFoundE
       enddo
       grid = ESMF_GridCreateCubedSphere(tileSize=tilesize, &
                                         coordSys=ESMF_COORDSYS_SPH_RAD, &
+                                        coordTypeKind=grid_typekind, &
                                         regDecompPTile=decomptile, &
                                         decompflagPTile=decompflagPTile, &
                                         name="fcst_grid", rc=rc)
@@ -215,6 +225,7 @@ if (rc /= ESMF_SUCCESS) write(0,*) 'rc=',rc,__FILE__,__LINE__; if(ESMF_LogFoundE
                                       maxIndex=(/nx,ny/), &
                                       gridAlign=(/-1,-1/), &
                                       coordSys=ESMF_COORDSYS_SPH_RAD, &
+                                      coordTypeKind=grid_typekind, &
                                       decompflag=(/ESMF_DECOMP_SYMMEDGEMAX,ESMF_DECOMP_SYMMEDGEMAX/), &
                                       name="fcst_grid", &
                                       indexflag=ESMF_INDEX_DELOCAL, &
```
When I run rap_control tests model passes the grid creation in fcst grid comp, but fails somewhere in physics, with this error:
```  0:   kgds(13-22)= -1431655766 -1431655766 -1431655766 -1431655766 -1431655766                                        
  0:  -1431655766           0         255 -1431655766 -1431655766                                                      
  0:  lat/lon grid                                                                                                     
  0:  imax,jmax,ijmax,dlon,dlat,ijordr,wlon,rnlat=                                                                     
  0:          180          91       16380   2.00000000000000                                                           
  0:   -2.00000000000000      T  0.000000000000000E+000   90.0000076293945                                             
  0:   FATAL ERROR: la2ga called                                                                                       
  0:   with lmask=true. But bad rslmsk                                                                                 
  0:   or slmask given.

DusanJovic-NOAA commented:
There are still type/kind mismatches in physics. For example, arrays in gcycle like xlon_d/xlat_d are declared as kind_phys (4 bytes now) while corresponding dummy arrays in sfccycle like rlo, rla are still declared as kind_io8 (8 bytes). There are many other mismatches.

My suggestions:
- get rid of all kinds in ccpp physics except:
         `kind_sngl_prec` fixed to 4 bytes
         `kind_dbl_prec` fixed to 8 bytes
         `kind_phys`, working precision set to either kind_sngl_prec or kind_dbl_prec at build time depending on the build option
- update (almost) all declarations to use `kind_phys`, except in very few cases where variable must be either 4 or 8 bytes all the time, like an I/O arrays in sfccycle or when interfacing with library routines that require specific kind (w3nco or sp).
- convert all files to be modules, no standalone routines anymore, like sfccycle
- get rid of compiler flags that ignore type mismatches

Time for some serious cleanup.

DusanJovic-NOAA commented:
Quick and dirty changes just to make the `rap_control` test 'work'. By 'work' I mean it executes without crashing.

See run directory: /scratch1/NCEPDEV/stmp2/Dusan.Jovic/rt_rap/rap_control

```diff
diff --git a/physics/machine.F b/physics/machine.F
index 9b09d235..ac881bc6 100644
--- a/physics/machine.F
+++ b/physics/machine.F
@@ -23,7 +23,7 @@
      &,                     kind_INTEGER = 4                             ! -,,-
 
 #else
-      integer, parameter :: kind_io4  = 4, kind_io8  = 8 , kind_ior = 8 &
+      integer, parameter :: kind_io4  = 4, kind_io8  = 4 , kind_ior = 8 &
      &,                     kind_evod = 4, kind_dbl_prec = 8            &
      &,                     kind_sngl_prec = 4
 # ifdef __PGI
diff --git a/physics/sfcsub.F b/physics/sfcsub.F
index 78e5201b..fbd7f49a 100644
--- a/physics/sfcsub.F
+++ b/physics/sfcsub.F
@@ -2736,7 +2736,7 @@
 !>\ingroup mod_sfcsub
       subroutine fixrdg(lugb,idim,jdim,fngrib,                          &
      &                  kpds5,gdata,gaus,blno,blto,me)
-      use machine      , only : kind_io8,kind_io4
+      use machine      , only : kind_io8,kind_dbl_prec,kind_sngl_prec
       use sfccyc_module, only : mdata
       implicit none
       integer lgrib,n,lskip,jret,j,ndata,lugi,jdim,idim,lugb,
@@ -2747,8 +2747,8 @@
       real (kind=kind_io8) gdata(idim*jdim)
       logical gaus
       real (kind=kind_io8) blno,blto
-      real (kind=kind_io8), allocatable :: data8(:)
-      real (kind=kind_io4), allocatable :: data4(:)
+      real (kind=kind_dbl_prec), allocatable :: data8(:)
+      real (kind=kind_sngl_prec), allocatable :: data4(:)
 !
       logical*1, allocatable :: lbms(:)
 !
@@ -2815,7 +2815,7 @@
         allocate(data4(1:idim*jdim))
         call getgb(lugb,lugi,kdata,lskip,jpds,jgds,ndata,lskip,
      &             kpds,kgds,lbms,data4,jret)
-        data8 = real(data4, kind=kind_io8)
+        data8 = real(data4, kind=kind_dbl_prec)
         deallocate(data4)
       else
         write(0,*)' FATAL ERROR: Invalid w3kindreal'
@@ -6165,7 +6165,7 @@
       subroutine setrmsk(kpds5,slmask,igaul,jgaul,wlon,rnlat,           &
      &                   data,imax,jmax,rlnout,rltout,lmask,rslmsk      &
      &,                  gaus,blno, blto, kgds1, kpds4, lbms)
-      use machine , only : kind_io8,kind_io4
+      use machine , only : kind_io8,kind_io4,kind_dbl_prec
       use sfccyc_module
       implicit none
       real (kind=kind_io8) blno,blto,wlon,rnlat,crit,data_max
@@ -6177,7 +6177,8 @@
       real (kind=kind_io8)    slmask(igaul,jgaul)
       real (kind=kind_io8)    data(imax,jmax),rslmsk(imax,jmax)
      &,                       rlnout(imax), rltout(jmax)
-      real (kind=kind_io8)    a(jmax), w(jmax), radi, dlat, dlon
+      real (kind=kind_io8)    radi, dlat, dlon
+      real (kind=kind_dbl_prec) a(jmax), w(jmax)
       logical lmask, gaus
 !
 !     set the longitude and latitudes for the grib file
@@ -6650,7 +6651,7 @@
 !! This subroutine interpolates from lat/lon grid to other lat/lon grid.
       subroutine ga2la(gauin,imxin,jmxin,regout,imxout,jmxout,          &
      &                 wlon,rnlat,rlnout,rltout,gaus,blno, blto)
-      use machine , only : kind_io8,kind_io4
+      use machine , only : kind_io8,kind_io4,kind_dbl_prec
       use sfccyc_module , only : num_threads
       implicit none
       integer i1,i2,j2,ishft,i,jj,j1,jtem,jmxout,imxin,jmxin,imxout,    &
@@ -6673,7 +6674,7 @@
       data    jmxsav/0/
       save    jmxsav, gaul, dlati
       real (kind=kind_io8) radi
-      real (kind=kind_io8) a(jmxin), w(jmxin)
+      real (kind=kind_dbl_prec) a(jmxin), w(jmxin)
 !
 !
       logical first
@@ -6917,11 +6918,12 @@ cjfe
 !>\ingroup mod_sfcsub
       subroutine gaulat(gaul,k)
 !
-      use machine , only : kind_io8,kind_io4
+      use machine , only : kind_io8,kind_io4,kind_dbl_prec
       implicit none
       integer n,k
       real (kind=kind_io8) radi
-      real (kind=kind_io8) a(k), w(k), gaul(k)
+      real (kind=kind_io8) gaul(k)
+      real (kind=kind_dbl_prec) a(k), w(k)
 !
       call splat(4, k, a, w)
 !
@@ -8305,7 +8307,7 @@ cjfe
      &                 gdata,len,iret                                   &
      &,                imsk, jmsk, slmskh, gaus,blno, blto              &
      &,                outlat, outlon, me)
-      use machine ,      only : kind_io8,kind_io4
+      use machine ,      only : kind_io8,kind_dbl_prec,kind_sngl_prec
       use sfccyc_module, only : mdata
       implicit none
       integer imax,jmax,ijmax,i,j,n,jret,inttyp,iret,imsk,              &
@@ -8321,8 +8323,8 @@ cjfe
 !
       real (kind=kind_io8) gdata(len), slmask(len)
       real (kind=kind_io8), allocatable :: data(:,:), rslmsk(:,:)
-      real (kind=kind_io8), allocatable :: data8(:)
-      real (kind=kind_io4), allocatable :: data4(:)
+      real (kind=kind_dbl_prec), allocatable :: data8(:)
+      real (kind=kind_sngl_prec), allocatable :: data4(:)
       real (kind=kind_io8), allocatable :: rlngrb(:), rltgrb(:)
 !
       logical lmask, yr2kc, gaus, ijordr
@@ -8398,7 +8400,7 @@ cjfe
         allocate(data4(1:mdata))
         call getgb(lugb,lugi,mdata,lskip,jpds,jgds,ndata,lskip,
      &             kpds,kgds,lbms,data4,jret)
-        data8 = real(data4, kind=kind_io8)
+        data8 = real(data4, kind=kind_dbl_prec)
         deallocate(data4)
       endif
       if (me .eq. 0) write(6,*) ' input grib file dates=',
@@ -8479,7 +8481,7 @@ cjfe
      &                  iy,im,id,ih,fh,gdata,len,iret                   &
      &,                 imsk, jmsk, slmskh, gaus,blno, blto             &
      &,                 outlat, outlon, me)
-      use machine      , only : kind_io8,kind_io4
+      use machine      , only : kind_io8,kind_dbl_prec,kind_sngl_prec
       use sfccyc_module, only : mdata
       implicit none
       integer nrepmx,nvalid,imo,iyr,idy,jret,ihr,nrept,lskip,lugi,      &
@@ -8506,8 +8508,8 @@ cjfe
 !
       real (kind=kind_io8) gdata(len), slmask(len)
       real (kind=kind_io8), allocatable :: data(:,:),rslmsk(:,:)
-      real (kind=kind_io8), allocatable :: data8(:)
-      real (kind=kind_io4), allocatable :: data4(:)
+      real (kind=kind_dbl_prec), allocatable :: data8(:)
+      real (kind=kind_sngl_prec), allocatable :: data4(:)
       real (kind=kind_io8), allocatable :: rlngrb(:), rltgrb(:)
 !
       logical lmask, yr2kc, gaus, ijordr
@@ -8645,7 +8647,7 @@ cjfe
         allocate (data4(1:mdata))
         call getgb(lugb,lugi,mdata,lskip,jpds,jgds,ndata,lskip,
      &             kpds,kgds,lbms,data4,jret)
-        data8 = real(data4, kind=kind_io8)
+        data8 = real(data4, kind=kind_dbl_prec)
         deallocate(data4)
       endif
       if (me .eq. 0) write(6,*) ' input grib file dates=',
```

SamuelTrahanNOAA commented:
I'm testing Dusan's changes against the regression tests. Wish me luck.

The next step is to get FV3_HRRR working so I can give this to the RRFS_dev1 people for testing.

DusanJovic-NOAA commented:
There are still few inconsistencies in how real arrays are passed to w3nco subroutines. ccpp-physics depends on w3nco_d (double precision version). Now that most of the real variables are either 4 or 8 bytes wide, each w3 call that has real arguments must either pass 8-byte reals, or detect which version of the library it's been linked against and use the correct arguments, which is done at most places, but not everywhere. For example in GFS_time_vary_pre.scm.F90, GFS_time_vary_pre.fv3.F90 it's currently passing 4-byte real to w3nco_d, which is wrong.

Since most of the w3 routines used in ccpp are dealing either with date/time computations or reading grib1 data, the precision of real arguments is not critical, in my opinion the best/simplest solution is to always link with w3nco_4 version and hard-code all real arguments to 4-byte kind. The alternative approach currently used (detecting kind of real values in w3nco) works, but produces argument mismatch warnings with GNU v10+ and forces us to use compiler flag to ignore all mismatches, which is bad.

SamuelTrahanNOAA commented:
This definitely doesn't work yet. When I run in debug mode, there are a ton of errors. Thus far, they have all been related to w3.

DusanJovic-NOAA commented:
See these two PRs:

https://github.com/SamuelTrahanNOAA/fv3atm/pull/3
https://github.com/SamuelTrahanNOAA/ccpp-physics/pull/4

SamuelTrahanNOAA commented:
This branch does seem to work with FV3_RAP and FV3_HRRR, but I want to test it with stochastic physics and a coupled case before merging it.

SamuelTrahanNOAA commented:
The 32bit.conf and 32bit_gnu.conf files should be removed before this is merged. They're in the PR now, just so the testers know what tests need new baselines.

junwang-noaa commented on tests/32bit.conf:
```diff
@@ -0,0 +1,19 @@
+COMPILE | -DAPP=ATM -DCCPP_SUITES=FV3_RAP,FV3_HRRR -D32BIT=ON -DCCPP_32BIT=ON                                                     |                | fv3 |
```
Since these tests are added in rt.conf/rt_gnu.conf, would you please remove the two 32bit.conf and 32bit_gnu.conf?

SamuelTrahanNOAA replied:
Those files are for the benefit of the testers who need to know which tests need new base lines.  It is *critical* that they do not regenerate baselines for all tests. We need to run the entire suite of 64-bit regression tests on all platforms to ensure the 64-bit functionality does not break, in any way.

SamuelTrahanNOAA replied:
If you read my earlier comment, I explain those files:

> The 32bit.conf and 32bit_gnu.conf files should be removed before this is merged. They're in the PR now, just so the testers know what tests need new baselines.

SamuelTrahanNOAA replied:
I've deleted those files.

junwang-noaa commented on tests/default_vars.sh:
```diff
@@ -270,6 +270,12 @@ fi
 
 WLCLK_dflt=30
 
+if [[ $MACHINE_ID = jet.* ]]; then
```
Are the wall clock time changes caused by the code changes from this PR?

SamuelTrahanNOAA replied:
My earlier PR added them, but somehow they were removed. It was necessary to get the 2threads tests to run in the wallclock limit. I will test to see if they're still necessary.

SamuelTrahanNOAA replied:
Some of the 32-bit physics, 64-bit dynamics tests take too long on hera.gnu. I will shorten their forecast lengths to compensate so we can use a 30 minute wallclock limit.

SamuelTrahanNOAA replied:
This is corrected now.

SamuelTrahanNOAA commented:
I just merged the latest changes, but it fails to build because the call to atmosphere_domain in atmos_model.F90 (fv3atm) is missing an argument (rd_domain) required by the call to atmosphere_domain in atmos_model.F90 (GFDL_atmos_cubed_sphere).

Is the authoritative repo in the middle of an update?

SamuelTrahanNOAA commented:
The atmos cubed sphere dev/emc is definitely ahead of the NOAA-EMC fv3atm

SamuelTrahanNOAA commented:
I reverted to the version of atmos_cubed_sphere used by NOAA-EMC fv3atm develop, and my branch passes regression tests on hera.intel, hera.gnu, and jet.intel.

SamuelTrahanNOAA commented:
I updated my branch again, and it appears the atmos_cubed_sphere is no longer ahead of FV3. I fixed two new type mismatches ("r*8" vs. "kind=kind_phys" in module_sf_noahmplsm) and everything works again on jet.intel, hera.intel, and hera.gnu.

SamuelTrahanNOAA commented:
Everything still works after merging #1292

SamuelTrahanNOAA commented:
The branch now works with the changes from [#1311](https://github.com/ufs-community/ufs-weather-model/pull/1311)

jkbk2004 commented:
@SamuelTrahanNOAA hera was extremely slow yesterday and today. We will merge in #1298 tomorrow morning for sure. Then we can start testing on this PR around tomorrow noon.

jkbk2004 commented:
@SamuelTrahanNOAA we can start testing. Baseline will change with this PR, right?

SamuelTrahanNOAA commented:
Don't start yet. I need to make sure it still works with the latest changes. It'll take another hour, if nothing goes wrong.

jkbk2004 commented:
> Don't start yet. I need to make sure it still works with the latest changes. It'll take another hour, if nothing goes wrong.

sure!

SamuelTrahanNOAA commented:
@jkbk2004 @BrianCurtis-NOAA This change will *not* change the baselines, but it will add new tests  It is *critical* for testing that you do not generate baselines for this. We need to confirm that the 64-bit physics will not change results for *any* test on *any* platform. 

For your convenience, the new tests are listed in `tests/32bit.conf` (intel) and `tests/32bit_gnu.conf` (gnu). They're also in `rt.conf` and `rt_gnu.conf`.

The `tests/32bit.conf` and `tests/32bit_gnu.conf` must be deleted before merge.

BrianCurtis-NOAA commented:
@SamuelTrahanNOAA OK. So would it work if we keep the current BL_DATE, add the 32 bit baselines to the current BL_DATE (since they are new tests, thats OK) and run the regular rt.conf/rt_gnu.conf? This eliminates the need for those 32_bit.conf files.

jkbk2004 commented:
> @jkbk2004 @BrianCurtis-NOAA This change will _not_ change the baselines, but it will add new tests It is _critical_ for testing that you do not generate baselines for this. We need to confirm that the 64-bit physics will not change results for _any_ test on _any_ platform.
> 
> For your convenience, the new tests are listed in `tests/32bit.conf` (intel) and `tests/32bit_gnu.conf` (gnu). They're also in `rt.conf` and `rt_gnu.conf`.
> 
> The `tests/32bit.conf` and `tests/32bit_gnu.conf` must be deleted before merge.

Yeah, I will go head to create and add those new additional BLs and run with current BL_DATE. Let me know when the PR is ready to start tests.

SamuelTrahanNOAA commented:
You forgot to copy the Jet baseline again.

jkbk2004 commented:
> You forgot to copy the Jet baseline again.

OMG! I am sorry. multiple meetings this morning!

BrianCurtis-NOAA commented:
> > You forgot to copy the Jet baseline again.
> 
> OMG! I am sorry. multiple meetings this morning!

Jong. Please make sure to have the baselines added to the BL_DIR _before_ running RT's on them (for all machines). This will make sure they are there. This is the 2nd time this has happened. If you message me off-hours I will move baselines on Jet ASAP.

jkbk2004 commented:
> > > You forgot to copy the Jet baseline again.
> > 
> > 
> > OMG! I am sorry. multiple meetings this morning!
> 
> Jong. Please make sure to have the baselines added to the BL_DIR _before_ running RT's on them (for all machines). This will make sure they are there. This is the 2nd time this has happened. If you message me off-hours I will move baselines on Jet ASAP.

Sure!

SamuelTrahanNOAA commented:
Some of the regional tests are failing for me with the Intel compiler on Hera. The model output matches, but the UPP output does not. This was not the case before I merged #1298 

I don't see any changes I made that could have caused this.

jkbk2004 commented:
> Some of the regional tests are failing for me with the Intel compiler on Hera. The model output matches, but the UPP output does not. This was not the case before I merged #1298
> 
> I don't see any changes I made that could have caused this.

So, I am wondering if some change is needed to follow on 32bit option?

jkbk2004 commented:
> > Some of the regional tests are failing for me with the Intel compiler on Hera. The model output matches, but the UPP output does not. This was not the case before I merged #1298
> > I don't see any changes I made that could have caused this.
> 
> So, I am wondering if some change is needed to follow on 32bit option?

On upp side as well?

SamuelTrahanNOAA commented:
The only possible cause I can see is that there was an error in #1298, where the code used the wrong data type in a call to a w3 or g2 routine. @DusanJovic-NOAA's changes will have corrected that since he explicitly queries the real kind before calling those routines. I do notice a comment about library changes in #1298.

Maybe @DusanJovic-NOAA can comment?

SamuelTrahanNOAA commented:
@jkbk2004 I should clarify. No tests are aborting. All run to completion and produce the correct model output. The GRIB2 files differ between some of the 64-bit tests and their baseline. I don't see any changes in my PR to how UPP is called.

DusanJovic-NOAA commented:
> The only possible cause I can see is that there was an error in #1298, where the code used the wrong data type in a call to a w3 or g2 routine. @DusanJovic-NOAA's changes will have corrected that since he explicitly queries the real kind before calling those routines. I do notice a comment about library changes in #1298.
> 
> Maybe @DusanJovic-NOAA can comment?

I do not see any library change in #1298

SamuelTrahanNOAA commented:
I'm out of ideas then.

jkbk2004 commented:
> > The only possible cause I can see is that there was an error in #1298, where the code used the wrong data type in a call to a w3 or g2 routine. @DusanJovic-NOAA's changes will have corrected that since he explicitly queries the real kind before calling those routines. I do notice a comment about library changes in #1298.
> > Maybe @DusanJovic-NOAA can comment?
> 
> I do not see any library change in #1298

upp side: it moved to w3emc2.9.2

DusanJovic-NOAA commented:
> > > The only possible cause I can see is that there was an error in #1298, where the code used the wrong data type in a call to a w3 or g2 routine. @DusanJovic-NOAA's changes will have corrected that since he explicitly queries the real kind before calling those routines. I do notice a comment about library changes in #1298.
> > > Maybe @DusanJovic-NOAA can comment?
> > 
> > 
> > I do not see any library change in #1298
> 
> upp side: it moved to w3emc2.9.2

Is that in standalone UPP. ufs-weather-model still depends on w3nco.

BrianCurtis-NOAA commented:
My PR (the last one) brought in UPP to it's HEAD of develop branch. The Hera tests passed in my PR, but that doesn't mean there isn't an issue.

SamuelTrahanNOAA commented:
The hera.gnu tests are passing. The only one left to run is rap_restart_dyn32_phy32, which is one of the new tests. Do the hera.gnu tests run UPP?

SamuelTrahanNOAA commented:
The other change is that w3nco was merged into w3emc in the post. Any of the w3 calls may have changed real kinds.

SamuelTrahanNOAA commented:
I think it is extremely likely this PR is an accidental bugfix for a real kind mismatch in #1298

DusanJovic-NOAA commented:
> The other change is that w3nco was merged into w3emc in the post. Any of the w3 calls may have changed real kinds.

ufs-weather-model (including upp library) still depends on w3nco, not w3emc. the change in standalone upp that updated w3nco to w3emc has no effect on the ufs-weather-model.

DusanJovic-NOAA commented:
> I think it is extremely likely this PR is an accidental bugfix for a real kind mismatch in #1298

Which real kind mismatch bug in #1298?

DusanJovic-NOAA commented:
> The hera.gnu tests are passing. The only one left to run is rap_restart_dyn32_phy32, which is one of the new tests. Do the hera.gnu tests run UPP?

gnu tests run inline_post (upp)

SamuelTrahanNOAA commented:
If the real kind of the library changed from, say, 4 to 8, then the calls that expected the old real kind would have failed. Your changes query the real kind used internally in the library, so the calls will be right regardless. The problem can vary by compiler if the libraries were built differently for each compiler (or if the wrong module was selected for a compiler).

SamuelTrahanNOAA commented:
@jkbk2004 Can you run the tests on WCOSS2 and see what happens?

junwang-noaa commented:
@SamuelTrahanNOAA The UFS is still using w3nco when compiling UPP, Please see upp cmakefile [here](https://github.com/NOAA-EMC/UPP/blob/develop/CMakeLists.txt#L60).

Also in the upp [PR#528](https://github.com/NOAA-EMC/UPP/pull/528), there are no source code and baseline change. My understanding is that the w3 subroutines were added to w3emc without any change. @aerorahul @kgerheiser Can you confirm?

DusanJovic-NOAA commented:
> If the real kind of the library changed from, say, 4 to 8, then the calls that expected the old real kind would have failed. Your changes query the real kind used internally in the library, so the calls will be right regardless. The problem can vary by compiler if the libraries were built differently for each compiler (or if the wrong module was selected for a compiler).

If this is the case then how do you explain that until this last merge you were getting identical upp outputs. The last PR #1298 did not change the library from 4 to 8 byte reals, nor did it change the compiler used to build the libraries.

kgerheiser commented:
@junwang-noaa the w3 routines were combined with no change to source code.

SamuelTrahanNOAA commented:
> If this is the case then how do you explain that until this last merge you were getting identical upp outputs. The last PR https://github.com/ufs-community/ufs-weather-model/pull/1298 did not change the library from 4 to 8 byte reals, nor did it change the compiler used to build the libraries.

Yes, I see that now. I have no more ideas. My PR doesn't change anything related to the post. The only tests that are failing so far are some of the HAFS and regional tests.

aerorahul commented:
> @SamuelTrahanNOAA The UFS is still using w3nco when compiling UPP, Please see upp cmakefile [here](https://github.com/NOAA-EMC/UPP/blob/develop/CMakeLists.txt#L60).
> 
> Also in the upp [PR#528](https://github.com/NOAA-EMC/UPP/pull/528), there are no source code and baseline change. My understanding is that the w3 subroutines were added to w3emc without any change. @aerorahul @kgerheiser Can you confirm?

@junwang-noaa I would check what hash of the UPP is being used in the UFS and if that hash still contains w3nco references.

BrianCurtis-NOAA commented:
> I would check what hash of the UPP is being used in the UFS and if that hash still contains w3nco references.

https://github.com/NOAA-EMC/UPP/tree/develop hash of [8767929](https://github.com/NOAA-EMC/UPP/tree/87679294301b56b9d1fe645b3eafa2a666f93870)

SamuelTrahanNOAA commented:
These tests failed on hera.intel: regional_control regional_2threads regional_control_2dwrtdecomp regional_3km regional_3km_decomp regional_spp_sppt_shum_skeb hafs_regional_atm

Nothing failed with hera.gnu. My jet.intel just started submitting test jobs (all compile jobs succeeded).

jkbk2004 commented:
> These tests failed on hera.intel: regional_control regional_2threads regional_control_2dwrtdecomp regional_3km regional_3km_decomp regional_spp_sppt_shum_skeb hafs_regional_atm
> 
> Nothing failed with hera.gnu. My jet.intel just started submitting test jobs (all compile jobs succeeded).

@SamuelTrahanNOAA can you point me to your experiment path on hera?

SamuelTrahanNOAA commented:
> can you point me to your experiment path on hera?

/scratch2/BMC/wrfruc/Samuel.Trahan/32bit/intel_single_prec_from_main
/scratch2/BMC/wrfruc/Samuel.Trahan/32bit/gnu_single_prec_from_main

The gnu tests have deleted their run directory, but the intel tests are waiting for me to magically fix the problem, so they still reside here:

/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_222762/

SamuelTrahanNOAA commented:
They're using these baselines, which are all symbolic links to the 20220713 directory, except for my new tests:

/scratch1/NCEPDEV/stmp4/Samuel.Trahan/FV3_RT/REGRESSION_TEST_INTEL

junwang-noaa commented:
@SamuelTrahanNOAA This reminds me that the upp updates are to fix issues with RRFS and HAFS. Can you double check if you merged with the latest fv3 updates (https://github.com/NOAA-EMC/fv3atm/commit/b9d61f2f0c51f834d5be26d5fbbd581f6d4e63d2)?

SamuelTrahanNOAA commented:
> @SamuelTrahanNOAA This reminds me that the upp updates are to fix issues with RRFS and HAFS. Can you double check if you merged with the latest fv3 updates ([NOAA-EMC/fv3atm@b9d61f2](https://github.com/NOAA-EMC/fv3atm/commit/b9d61f2f0c51f834d5be26d5fbbd581f6d4e63d2))?

I've already checked seven times over the last hour, and all seven times I saw that my branches are up-to-date, and I point to the same UPP as develop.

SamuelTrahanNOAA commented:
What truly baffles me is this: I cannot think of anything those tests have in common, which they don't have in common with other regional or HAFS tests.

BrianCurtis-NOAA commented:
If you isolate the failed intel tests and run them with current develop of ufs-weather-model on Hera, do the same issues arise?

SamuelTrahanNOAA commented:
> If you isolate the failed intel tests and run them with current develop of ufs-weather-model on Hera, do the same issues arise?

No. I just tried that. All of those tests passed off of the top of ufs-community develop.

I just started another attempt, running the failed tests in my branch against the emc.nemspara baselines instead of my symbolic links to them.

jkbk2004 commented:
@BrianCurtis-NOAA I have a different question. If I check intel BL directories, I don't see regional_control.

BrianCurtis-NOAA commented:
> @BrianCurtis-NOAA I have a different question. If I check intel BL directories, I don't see regional_control.

Which machines?

SamuelTrahanNOAA commented:
@jkbk2004 It is fv3_regional_control

```
###############################################################################
#
#  FV3 regional control test
#
###############################################################################

export TEST_DESCR="Compare FV3 regional results with previous trunk version"

export CNTL_DIR=fv3_regional_control

```

jkbk2004 commented:
> @jkbk2004 It is fv3_regional_control
> 
> ```
> ###############################################################################
> #
> #  FV3 regional control test
> #
> ###############################################################################
> 
> export TEST_DESCR="Compare FV3 regional results with previous trunk version"
> 
> export CNTL_DIR=fv3_regional_control
> ```

Got it!

jkbk2004 commented:
> > @jkbk2004 It is fv3_regional_control
> > ```
> > ###############################################################################
> > #
> > #  FV3 regional control test
> > #
> > ###############################################################################
> > 
> > export TEST_DESCR="Compare FV3 regional results with previous trunk version"
> > 
> > export CNTL_DIR=fv3_regional_control
> > ```
> 
> Got it!

@SamuelTrahanNOAA why issue is only with those regional cases?

SamuelTrahanNOAA commented:
@jkbk2004 I just reran the failed test cases off of the emc.nemspara baselines, and they succeeded this time. Maybe the problem was with my personal baseline directory? Perhaps symlinks don't always work as expected?

Could you please try running the hera.intel tests off of this branch?

jkbk2004 commented:
> @jkbk2004 I just reran the failed test cases off of the emc.nemspara baselines, and they succeeded this time. Maybe the problem was with my personal baseline directory? Perhaps symlinks don't always work as expected?
> 
> Could you please try running the hera.intel tests off of this branch?

sure! I will keep you posted. I may try on cheyenne as well. sounds like getting into test mode.

SamuelTrahanNOAA commented:
@jkbk2004 Have you had any success? I see the wcoss2.intel tests passed at least.

jkbk2004 commented:
hera/nems account low fairshare issue. I am working on orion. give me an hour or two. On Cheyenne, we got hit with max core hour issue as well. Once orion passes, I will keep moving to jet/gaea.

jkbk2004 commented:
@SamuelTrahanNOAA it passes on orion. I will move on to gaea and jet.

jkbk2004 commented:
@SamuelTrahanNOAA Cheyenne is not available due to account core hour resource. Tests passed on all other machines. @DusanJovic-NOAA @junwang-noaa @DeniseWorthen @ChunxiZhang-NOAA We can move on for final review and start merging process.

DeniseWorthen commented:
@jkbk2004 Where did the core hours go on Cheyenne? We've never run out of core hours on Cheyenne that I know of.

jkbk2004 commented:
> @jkbk2004 Where did the core hours go on Cheyenne? We've never run out of core hours on Cheyenne that I know of.
With current scale of RT, we need about 2 mil core hours. current core hour 1 mil maxed out (half of the year). Plus ufs-wm .vs. srw cost ratio is about 2:1.

jkbk2004 commented:
> @jkbk2004 Where did the core hours go on Cheyenne? We've never run out of core hours on Cheyenne that I know of.

2 mil core hour per year

jkbk2004 commented:
> > @jkbk2004 Where did the core hours go on Cheyenne? We've never run out of core hours on Cheyenne that I know of.
> 
> 2 mil core hour per year

Last April, I reported that we need about 1.8 mil last April. But it sounds like we need 2 mil at least.

DeniseWorthen commented:
So our allocation on Cheyenne is used for not just regression testing, but for some sort of SRW verification?

jkbk2004 commented:
> So our allocation on Cheyenne is used for not just regression testing, but for some sort of SRW verification?



> So our allocation on Cheyenne is used for not just regression testing, but for some sort of SRW verification?

Yes, SCSG0* account is a shared account across groups, I guess.

SamuelTrahanNOAA commented:
Does this mean Cheyenne is no longer a supported platform, until you get resources again?

jkbk2004 commented:
> Does this mean Cheyenne is no longer a supported platform, until you get resources again?

Yes, we are not able to support Cheyenne until we get additional resource.

climbfuji commented:
I think it is unwise to drop Cheyenne from the Tier-1 list, even if only temporary. We all know what happens if a system isn't tested for a few weeks, it mostly means that the code stops working. Can we switch back to the RAL allocation until the problems are resolved? @mkavulich @ligiabernardet @junwang-noaa ?

SamuelTrahanNOAA commented:
I do want this PR tested on Cheyenne before it is merged. Cheyenne has an opinionated compiler, so it is the best platform for finding some types of coding issues.

ligiabernardet commented:
I agree that we need to test on Cheyenne, especially for a complex PR such as this one that touches many parts of the code. I cannot speak for using RAL resources but I will reach out to Mike Ek.

jkbk2004 commented:
> I do want this PR tested on Cheyenne before it is merged. Cheyenne has an opinionated compiler, so it is the best platform for finding some types of coding issues.
@ligiabernardet Is there a way to add me to RAL account even if temporary? that might be quickest approach.

jkbk2004 commented:
> > I do want this PR tested on Cheyenne before it is merged. Cheyenne has an opinionated compiler, so it is the best platform for finding some types of coding issues.
> > @ligiabernardet Is there a way to add me to RAL account even if temporary? that might be quickest approach.

So nice to see that CISL help desk is adding 0.5 mil hours to our current (SCSG0*). I will check.

jkbk2004 commented:
RTs are running again on Cheyenne. @DomHeinzeller @ligiabernardet @mkavulich Thanks!

jkbk2004 commented:
@SamuelTrahanNOAA Thanks to cisl's quick backup for adding resources, Cheyenne tests are done. Lets start merging in dependencies. The ccpp pr seems to be ready.

SamuelTrahanNOAA commented:
I have merged the CCPP PR, but we need to wait until https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/193 is merged before we can merge the https://github.com/NOAA-EMC/fv3atm/pull/534

@jkbk2004 You should post a message at https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/193 asking them to merge it.

jkbk2004 commented:
> I have merged the CCPP PR, but we need to wait until [NOAA-GFDL/GFDL_atmos_cubed_sphere#193](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/193) is merged before we can merge the [NOAA-EMC/fv3atm#534](https://github.com/NOAA-EMC/fv3atm/pull/534)
> 
> @jkbk2004 You should post a message at [NOAA-GFDL/GFDL_atmos_cubed_sphere#193](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/193) asking them to merge it.

Sure! I will go to the PR there, [stochastic pr59](https://github.com/noaa-psd/stochastic_physics/pull/59) as well.

SamuelTrahanNOAA commented:
It appears the https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/193 PR is waiting for a review from @junwang-noaa and @DomHeinzeller, requested by the repository's code managers.

DeniseWorthen commented:
@jkbk2004 The log for Hera.gnu was run against a local baseline generated by Sam. Do we know for sure that the additional baselines for GNU have been added to the current develop baseline on Hera?

SamuelTrahanNOAA commented:
@jkbk2004 I'd like someone else to independently run all tests. If you haven't run the hera.gnu tests, could you please do so? Those tests don't take long since they're few in number and don't wait for Intel compiler licenses.

jkbk2004 commented:
@SamuelTrahanNOAA let me double check hera-gnu

SamuelTrahanNOAA commented:
The chances of the hera.gnu tests failing are extraordinarily low since I've tested those many times myself. However, I would like independent testing since this PR has such broad implications.

BrianCurtis-NOAA commented:
> The chances of the hera.gnu tests failing are extraordinarily low since I've tested those many times myself. However, I would like independent testing since this PR has such broad implications.

Yes, they should always be run by EPIC separate from the owner of the PR. It's very helpful to know what machines the PR owners tests on as well.

SamuelTrahanNOAA commented:
This branch now points to the authoritative repositories for FV3 and stochastic_physics. It is ready for final review and merge.

SamuelTrahanNOAA commented:
Hold on. I need to delete two files first.

SamuelTrahanNOAA commented:
Okay. NOW the branch is ready for final review and merge.

SamuelTrahanNOAA commented:
I need one more review before this can be merged.

jkbk2004 commented:
@DusanJovic-NOAA @DeniseWorthen @ChunxiZhang-NOAA Finally, everything is ready to get final review and approval.

DusanJovic-NOAA commented:
fv3atm submodule hash needs to be https://github.com/NOAA-EMC/fv3atm/commit/9743346431c46642958712690e2c2733763ce5de

BrianCurtis-NOAA commented:
> fv3atm submodule hash needs to be [NOAA-EMC/fv3atm@9743346](https://github.com/NOAA-EMC/fv3atm/commit/9743346431c46642958712690e2c2733763ce5de)

@SamuelTrahanNOAA ^^

SamuelTrahanNOAA commented:
@BrianCurtis-NOAA I just updated the FV3 hash.
